internal/controlplane/handlers_authz.go

```diff
@@ -292,32 +296,157 @@ func (s *Server) ListRoleAssignments(
 // AssignRole assigns a role to a user on a project.
 // Note that this assumes that the request has already been authorized.
 func (s *Server) AssignRole(ctx context.Context, req *minder.AssignRoleRequest) (*minder.AssignRoleResponse, error) {
-	// Request Validation
 	role := req.GetRoleAssignment().GetRole()
 	sub := req.GetRoleAssignment().GetSubject()
+	email := req.GetRoleAssignment().GetEmail()
 
-	if role == "" || sub == "" {
-		return nil, util.UserVisibleError(codes.InvalidArgument, "role and subject must be specified")
+	// Determine the target project.
+	entityCtx := engine.EntityFromContext(ctx)
+	targetProject := entityCtx.Project.ID
+
+	// Ensure user is not updating their own role
+	err := s.isUserSelfUpdating(ctx, sub, email)
+	if err != nil {
+		return nil, err
 	}
 
 	// Parse role (this also validates)
-	authzrole, err := authz.ParseRole(role)
+	authzRole, err := authz.ParseRole(role)
 	if err != nil {
 		return nil, util.UserVisibleError(codes.InvalidArgument, err.Error())
 	}
 
+	// Ensure the target project exists
+	_, err = s.store.GetProjectByID(ctx, targetProject)
+	if err != nil {
+		// If the project is not found, return an error
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.InvalidArgument, "target project with ID %s not found", targetProject)
+		}
+		return nil, status.Errorf(codes.Internal, "error getting project: %v", err)
+	}
+
+	// Validate the subject and email - decide if it's an invitation or a role assignment
+	if sub == "" && email != "" && isEmail(email) {
```

Comment:
if `isEmail()` returns false here I think we should return an internal error as it points to an internal data inconsistency already present in the database.

Reply:
The email in this case is received as an argument that is part of the request message, so I think it makes sense to validate it and return `codes.InvalidArgument` if it doesn't look alright